### PR TITLE
VQAv2 finetune eval

### PR DESCRIPTION
**Pull request by andrewwangva**

### CODE REVIEW

File: src/training/data.py [hunk 1]
```diff
@@ -73,8 +72,8 @@ def set_epoch(self, epoch):
 
 
 def get_dataset_size(shards):
-    shards_list = list(braceexpand.braceexpand(shards))
-    dir_path = os.path.dirname(shards)
+    shards_list = wds.shardlists.expand_urls(shards)
```
Comment (gmittal) [inline src/training/data.py:75]: ???

File: src/training/vqa_fine_tune.py [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+from training.scheduler import cosine_lr
+from training.train import AverageMeter
+import evaluate
+from sklearn import preprocessing
+import numpy as np
+
+from datasets import load_dataset_builder
+from datasets import load_dataset
+
+class VQATextDataset(Dataset):
```
Comment (gmittal) [inline src/training/vqa_fine_tune.py:23]: nit: `VQAv2Dataset`

File: src/training/vqa_fine_tune.py [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+            #references=label.cpu().numpy(),
+        #)
+    model.train()
+    #metrics = metric.compute()
+    metrics = {}
+    metrics["loss"] = val_loss / samples_seen
+    
+    return metrics
+
+def train_single_epoch(model, data, optimizer, args):
```
Comment (gmittal) [inline src/training/vqa_fine_tune.py:174]: why is there both a `train_one_epoch` and `train_single_epoch`?

File: src/training/vqa_fine_tune.py [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+        optimizer.step()
+
+        progress_bar.set_description(f"Loss: {loss.item():.4f}")
+        progress_bar.update(1)
+        
+        if (i % args.val_frequency) == 0 and i > 0:
+            print(loss)
+            metrics = compute_metrics(model, data["train"], device, args)
+            
+            #end_training = early_stop.step(metrics)
```
Comment (gmittal) [inline src/training/vqa_fine_tune.py]: not checking for early stopping?

File: src/training/vqa_fine_tune.py [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+        help="Use a pretrained CLIP model weights with the specified tag or file path.",
+    )
+    parser.add_argument(
+        "--seed", type=int, default=0, help="Default random seed."
+    )
+
+    args = parser.parse_args(args)
+    return args
+
+dataset_train = load_dataset("HuggingFaceM4/VQAv2", split="train", cache_dir = "./sample_data")
```
Comment (gmittal) [inline src/training/vqa_fine_tune.py]: want to put this inside of a `def main(...)` function?

File: src/training/vqa_fine_tune.py [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+class_id = np.array(class_id)
+questions = np.array(questions)
+images = np.array(images)
+answers = np.array(answers)
+
+dataset_df = pd.DataFrame({'question_id': class_id, 'question': questions, 'image': images, 'multiple_choice_answer': answers})
+dataset_df = dataset_df[0:12800]
+
+
+args = parse_args([])
```
Comment (gmittal) [inline src/training/vqa_fine_tune.py]: why is empty list passed to parse_args?

File: src/training/vqa_fine_tune.py [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+
+    return dataloaders
+
+class CLIPMultimodalClassifier(nn.Module):
+    def __init__(self, encoder, embed_dim, num_labels):
+        super().__init__()
+
+        self.encoder = encoder
+        
+        self.fc1 = nn.Linear(embed_dim * 2, 1536) #size of answer space
```
Comment (gmittal) [inline src/training/vqa_fine_tune.py]: would recommend using `nn.Sequential` like this:

```python
self.mlp = nn.Sequential(
    nn.Linear(embed_dim * 2, 1536),
    nn.GELU(),
    nn.LayerNorm(1536),
    nn.Linear(1536, num_classes),
)
```

and then in the forward:

```
self.mlp(...)
```
rather than doing the whole `self.lnorm(F.relu(...)))`

File: src/training/vqa_fine_tune.py [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+            image = batch["image"].to(device)
+            text = batch["text"].to(device)
+            label = batch["label"].to(device)
+            samples_seen += text.shape[0]
+            logits = model(image, text)
+
+            #predictions = torch.argmax(logits)
+            batch_val_loss = loss_fn(logits, label)
+        val_loss += batch_val_loss.item()
+        #metric.add_batch(
```
Comment (gmittal) [inline src/training/vqa_fine_tune.py]: are you using accuracy as the metric? it doesn't seem that you're computing it here.

Comment (gmittal) [reply src/training/vqa_fine_tune.py]: how is the `test-dev VQA score` (as mentioned in the FLAVA paper) computed?

File: src/training/data.py (hunk 1 above)
Comment (andrewwangva) [reply src/training/data.py:75]: uhh I have no idea why this was changed whoops

Comment (gmittal) [reply src/training/data.py:75]: I just realized it's from the first couple of commits from the OpenCLIP authors. Don't worry about it.

File: src/training/vqa_fine_tune.py (hunk 3 above)
Comment (andrewwangva) [reply src/training/vqa_fine_tune.py:174]: I use train_single_epoch for testing purposes when I don't want anything besides the bare bones NN. It'll most likely be removed in future versions

File: src/training/vqa_fine_tune.py (hunk 5 above)
Comment (andrewwangva) [reply src/training/vqa_fine_tune.py]: yes, I will do that.

**Comment by gmittal** [comment]
@andrewwangva how is this different from #8?

File: src/training/vqa_fine_tune.py [hunk 9]
```diff
@@ -0,0 +1,338 @@
...
+    for split in ["train", "validation"]:
+        dataset_train = load_dataset(path, split=split, cache_dir = "./sample_data")
+        dataset_df = dataset_train.to_pandas()
+
+        class_id = []
+        questions = []
+        images = []
+        answers = []
+        weights = []
+        for index, row in dataset_df.iterrows():
```
Comment (gmittal) [inline src/training/vqa_fine_tune.py]: This is a lot of preprocessing, why is it not in the Dataset class directly?

File: src/training/vqa_fine_tune.py [hunk 10]
```diff
@@ -0,0 +1,338 @@
...
+
+        class_id = np.array(class_id)
+        questions = np.array(questions)
+        images = np.array(images)
+        dataset_df = pd.DataFrame({'question_id': class_id, 'question': questions, 'image': images, 'answer_list': answers, 'answer_weights': weights})
+        #dataset_df = dataset_df[0:12800]
+        b_size = args.batch_size
+        if(split == "validation"):
+            b_size = args.batch_size * 20
+            dataset_df = dataset_df[0:12800]
```
Comment (gmittal) [inline src/training/vqa_fine_tune.py:77]: it's fine if you have this for debugging, but you should leave a `# TODO: ...` so we know to remove it before merging

File: src/training/vqa_fine_tune.py [hunk 11]
```diff
@@ -0,0 +1,338 @@
...
+
+def get_score(count: int) -> float:
+    return min(1.0, count / 3)
+
+def get_task_dataloaders(path, transforms, labelencoder, answer_set, args):
+    tokenizer = get_tokenizer(args.model)
+    dataloaders = {}
+    
+    for split in ["train", "validation"]:
+        dataset_train = load_dataset(path, split=split, cache_dir = "./sample_data")
```
Comment (gmittal) [inline src/training/vqa_fine_tune.py]: This should really be done in the `VQATextDataset` init method

File: src/training/vqa_fine_tune.py [hunk 12]
```diff
@@ -0,0 +1,338 @@
...
+        patience=args.early_stop_patience,
+        threshold=args.early_stop_threshold,
+        metric_name=args.early_stop_metric_name,
+    )
+
+    for epoch in range(20):
+        val_metrics, end_training = train_one_epoch(clf, data, epoch, optim, scheduler, early_stop, device, args)
+
+if __name__ == "__main__":
+    main(sys.argv[1:])
```
Comment (gmittal) [inline src/training/vqa_fine_tune.py]: Add newline at end of file
